File: ta4j-core/src/main/java/org/ta4j/core/num/Num.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.num;

import java.io.Serializable;
import java.math.BigDecimal;
import java.math.MathContext;

/**
 * Ta4js definition of operations that must be fulfilled by an object that
 * should be used as base for calculations.
 *
 * @see Num
 * @see DoubleNum
 * @see DecimalNum
 */
public interface Num extends Comparable<Num>, Serializable {

    /**
     * @return the delegate used from this {@code Num} implementation
     */
    Number getDelegate();

    /**
     * @return factory that created this instance with defined precision
     */
    NumFactory getNumFactory();

    /**
     * Returns the name/description of this Num implementation.
     *
     * @return the name/description
     */
    String getName();

    /**
     * Returns a {@code Num} whose value is {@code (this + augend)}.
     *
     * @param augend value to be added to this {@code Num}
     * @return {@code this + augend}, rounded as necessary
     */
    Num plus(Num augend);

    /**
     * Returns a {@code Num} whose value is {@code (this - augend)}.
     *
     * @param subtrahend value to be subtracted from this {@code Num}
     * @return {@code this - subtrahend}, rounded as necessary
     */
    Num minus(Num subtrahend);

    /**
     * Returns a {@code Num} whose value is {@code this * multiplicand}.
     *
     * @param multiplicand value to be multiplied by this {@code Num}
     * @return {@code this * multiplicand}, rounded as necessary
     */
    Num multipliedBy(Num multiplicand);

    /**
     * Returns a {@code Num} whose value is {@code (this / divisor)}.
     *
     * @param divisor value by which this {@code Num} is to be divided
     * @return {@code this / divisor}, rounded as necessary
     */
    Num dividedBy(Num divisor);

    /**
     * Returns a {@code Num} whose value is {@code (this % divisor)}.
     *
     * @param divisor value by which this {@code Num} is to be divided
     * @return {@code this % divisor}, rounded as necessary
     */
    Num remainder(Num divisor);

    /**
     * Returns a {@code Num} whose value is rounded down to the nearest whole
     * number.
     *
     * @return {@code this} to whole Num rounded down
     */
    Num floor();

    /**
     * Returns a {@code Num} whose value is rounded up to the nearest whole number.
     *
     * @return {@code this} to whole Num rounded up
     */
    Num ceil();

    /**
     * Returns a {@code Num} whose value is <code>(this<sup>n</sup>)</code>.
     *
     * @param n power to raise this {@code Num} to.
     * @return <code>this<sup>n</sup></code>
     */
    Num pow(int n);

    /**
     * Returns a {@code Num} whose value is <code>(this<sup>n</sup>)</code>.
     *
     * @param n power to raise this {@code Num} to.
     * @return <code>this<sup>n</sup></code>
     */
    Num pow(Num n);

    /**
     * Returns a {@code Num} whose value is {@code log(this)}.
     *
     * @return {@code log(this)}
     */
    Num log();

    /**
     * Returns a {@code Num} whose value is {@code e^this}.
     *
     * @return {@code e^this}
     */
    Num exp();

    /**
     * Returns a {@code Num} whose value is {@code √(this)}.
     *
     * @return {@code √(this)}
     */
    Num sqrt();

    /**
     * Returns a {@code Num} whose value is {@code √(this)}.
     *
     * @param mathContext to calculate.
     * @return {@code √(this)}
     */
    Num sqrt(MathContext mathContext);

    /**
     * Returns a {@code Num} whose value is the absolute value of this {@code Num}.
     *
     * @return {@code abs(this)}
     */
    Num abs();

    /**
     * Returns a {@code Num} whose value is (-this), and whose scale is
     * this.scale().
     *
     * @return {@code negate(this)}
     */
    Num negate();

    /**
     * Checks if {@code this} is zero.
     *
     * @return true if {@code this == 0}, false otherwise
     */
    boolean isZero();

    /**
     * Checks if {@code this} is greater than zero.
     *
     * @return true if {@code this > 0}, false otherwise
     */
    boolean isPositive();

    /**
     * Checks if {@code this} is zero or greater.
     *
     * @return true if {@code this ≥ 0}, false otherwise
     */
    boolean isPositiveOrZero();

    /**
     * Checks if {@code this} is less than zero.
     *
     * @return true if {@code this < 0}, false otherwise
     */
    boolean isNegative();

    /**
     * Checks if {@code this} is zero or less.
     *
     * @return true if {@code this ≤ 0}, false otherwise
     */
    boolean isNegativeOrZero();

    /**
     * Checks if {@code this} is equal to {@code other}.
     *
     * @param other the other value, not null
     * @return true if {@code this == other}, false otherwise
     */
    boolean isEqual(Num other);

    /**
     * Checks if {@code this} is greater than {@code other}.
     *
     * @param other the other value, not null
     * @return true if {@code this > other}, false otherwise
     */
    boolean isGreaterThan(Num other);

    /**
     * Checks if {@code this} is greater than or equal to {@code other}.
     *
     * @param other the other value, not null
     * @return true if {@code this ≥ other}, false otherwise
     */
    boolean isGreaterThanOrEqual(Num other);

    /**
     * Checks if {@code this} is less than {@code other}.
     *
     * @param other the other value, not null
     * @return true if {@code this < other}, false otherwise
     */
    boolean isLessThan(Num other);

    /**
     * Checks if {@code this} is less than {@code other}.
     *
     * @param other the other value, not null
     * @return true if {@code this ≤ other}, false otherwise
     */
    boolean isLessThanOrEqual(Num other);

    /**
     * Returns the {@code Num} whose value is the smaller of {@code this} and
     * {@code other}.
     *
     * @param other value with which to calculate the minimum
     * @return the smaller of {@code this} and {@code other}. If they are equal,
     *         {@code this} is returned.
     */
    Num min(Num other);

    /**
     * Returns the {@code Num} whose value is the greater of {@code this} and
     * {@code other}.
     *
     * @param other value with which to calculate the maximum
     * @return the greater of {@code this} and {@code other}. If they are equal,
     *         {@code this} is returned.
     */
    Num max(Num other);

    /**
     * Returns true only if {@code this} is an instance of {@link NaN}.
     *
     * @return false if this implementation is not {@link NaN}
     */
    default boolean isNaN() {
        return false;
    }

    /**
     * Converts this {@code Num} to an {@code integer}.
     *
     * @return this {@code Num} converted to an {@code integer}
     */
    default int intValue() {
        return getDelegate().intValue();
    }

    /**
     * Converts this {@code Num} to a {@code long}.
     *
     * @return this {@code Num} converted to a {@code long}
     */
    default long longValue() {
        return getDelegate().longValue();
    }

    /**
     * Converts this {@code Num} to a {@code float}.
     *
     * @return this {@code Num} converted to a {@code float}
     */
    default float floatValue() {
        return getDelegate().floatValue();
    }

    /**
     * Converts this {@code Num} to a {@code double}.
     *
     * @return this {@code Num} converted to a {@code double}
     */
    default double doubleValue() {
        return getDelegate().doubleValue();
    }

    /**
     * Checks if a Num value is null or NaN.
     *
     * <p>
     * This method performs comprehensive NaN detection by checking:
     * <ul>
     * <li>If the value is null</li>
     * <li>If {@link Num#isNaN()} returns true</li>
     * <li>If the underlying double value is {@link Double#NaN} (handles DoubleNum
     * edge cases)</li>
     * </ul>
     *
     * <p>
     * This is necessary because {@link DoubleNumFactory} can surface
     * {@link Double#NaN} values that may not satisfy {@link Num#isNaN()}.
     *
     * @param value the value to check, may be null
     * @return true if the value is null or NaN, false otherwise
     */
    static boolean isNaNOrNull(Num value) {
        return value == null || value.isNaN() || Double.isNaN(value.doubleValue());
    }

    /**
     * Checks if a Num value is valid (not null and not NaN).
     *
     * <p>
     * This is the logical complement of {@link #isNaNOrNull(Num)}. A value is
     * considered valid if it is non-null and represents a real number (not NaN).
     *
     * @param value the value to check, may be null
     * @return true if the value is non-null and not NaN, false otherwise
     * @since 0.22.0
     */
    static boolean isValid(Num value) {
        return !isNaNOrNull(value);
    }

    /**
     * Converts this {@code Num} to a {@code BigDecimal}.
     *
     * @return this {@code Num} converted to a {@code BigDecimal}
     */
    BigDecimal bigDecimalValue();

    @Override
    int hashCode();

    @Override
    String toString();

    /**
     * {@inheritDoc}
     */
    @Override
    boolean equals(Object obj);

}

File: ta4j-core/src/main/kotlin/org/ta4j/core/num/Num.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.num

import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode
import java.util.function.Function

/**
 * Ta4js definition of operations that must be fulfilled by an object that
 * should be used as base for calculations
 *
 * @see Num
 *
 * @see Num.function
 * @see DoubleNum
 *
 * @see DecimalNum
 */
interface Num : Comparable<Num> {
    /**
     * @return the Num of 0
     */
    fun zero(): Num {
        return numOf(0)
    }

    /**
     * @return the Num of 1
     */
    fun one(): Num {
        return numOf(1)
    }

    /**
     * @return the Num of 100
     */
    fun hundred(): Num {
        return numOf(100)
    }

    /**
     * @return the delegate used from this `Num` implementation
     */
    val delegate: Number?

    /**
     * Returns the name/description of this Num implementation
     *
     * @return the name/description
     */
    val name: String

    /**
     * Returns a `num` whose value is `(this + augend)`,
     *
     * @param augend value to be added to this `num`.
     * @return `this + augend`, rounded as necessary
     */
    operator fun plus(augend: Num): Num

    /**
     * Returns a `num` whose value is `(this - augend)`,
     *
     * @param subtrahend value to be subtracted from this `num`.
     * @return `this - subtrahend`, rounded as necessary
     */
    operator fun minus(subtrahend: Num): Num

    /**
     * Returns a `num` whose value is `this * multiplicand`,
     *
     * @param multiplicand value to be multiplied by this `num`.
     * @return `this * multiplicand`, rounded as necessary
     */
    fun multipliedBy(multiplicand: Num): Num

    operator fun times(multiplicand: Num):Num =multipliedBy(multiplicand)

    /**
     * Returns a `num` whose value is `(this / divisor)`,
     *
     * @param divisor value by which this `num` is to be divided.
     * @return `this / divisor`, rounded as necessary
     */
    fun dividedBy(divisor: Num): Num

    operator fun div(divisor: Num): Num=dividedBy(divisor)

    /**
     * Returns a `num` whose value is `(this % divisor)`,
     *
     * @param divisor value by which this `num` is to be divided.
     * @return `this % divisor`, rounded as necessary.
     */
    fun remainder(divisor: Num): Num

    operator fun rem(divisor: Num): Num= remainder(divisor)

    /**
     * Returns a `Num` whose value is rounded down to the nearest whole
     * number.
     *
     * @return `this` to whole Num rounded down
     */
    fun floor(): Num

    /**
     * Returns a `Num` whose value is rounded up to the nearest whole number.
     *
     * @return `this` to whole Num rounded up
     */
    fun ceil(): Num

    /**
     * Returns a `num` whose value is `(this<sup>n</sup>)`.
     *
     * @param n power to raise this `num` to.
     * @return `this<sup>n</sup>`
     */
    fun pow(n: Int): Num

    /**
     * Returns a `num` whose value is `(this<sup>n</sup>)`.
     *
     * @param n power to raise this `num` to.
     * @return `this<sup>n</sup>`
     */
    fun pow(n: Num): Num

    /**
     * Returns a `num` whose value is `ln(this)`.
     *
     * @return `this<sup>n</sup>`
     */
    fun log(): Num

    /**
     * Returns a `num` whose value is `√(this)`.
     *
     * @return `this<sup>n</sup>`
     */
    fun sqrt(): Num

    /**
     * Returns a `num` whose value is `√(this)`.
     *
     * @param precision to calculate.
     * @return `this<sup>n</sup>`
     */
    fun sqrt(precision: Int): Num

    /**
     * Returns a `num` whose value is the absolute value of this `num`.
     *
     * @return `abs(this)`
     */
    fun abs(): Num

    /**
     * Returns a `num` whose value is (-this), and whose scale is
     * this.scale().
     *
     * @return `negate(this)`
     */
    fun negate(): Num

    /**
     * Checks if the value is zero.
     *
     * @return true if the value is zero, false otherwise
     */
    val isZero: Boolean

    /**
     * Checks if the value is greater than zero.
     *
     * @return true if the value is greater than zero, false otherwise
     */
    val isPositive: Boolean

    /**
     * Checks if the value is zero or greater.
     *
     * @return true if the value is zero or greater, false otherwise
     */
    val isPositiveOrZero: Boolean

    /**
     * Checks if the value is less than zero.
     *
     * @return true if the value is less than zero, false otherwise
     */
    val isNegative: Boolean

    /**
     * Checks if the value is zero or less.
     *
     * @return true if the value is zero or less, false otherwise
     */
    val isNegativeOrZero: Boolean

    /**
     * Checks if this value is equal to another.
     *
     * @param other the other value, not null
     * @return true if this is greater than the specified value, false otherwise
     */
    fun isEqual(other: Num?): Boolean

    /**
     * Checks if this value is greater than another.
     *
     * @param other the other value, not null
     * @return true if this is greater than the specified value, false otherwise
     */
    fun isGreaterThan(other: Num?): Boolean

    /**
     * Checks if this value is greater than or equal to another.
     *
     * @param other the other value, not null
     * @return true if this is greater than or equal to the specified value, false
     * otherwise
     */
    fun isGreaterThanOrEqual(other: Num?): Boolean

    /**
     * Checks if this value is less than another.
     *
     * @param other the other value, not null
     * @return true if this is less than the specified value, false otherwise
     */
    fun isLessThan(other: Num?): Boolean

    /**
     * Checks if this value is less than another.
     *
     * @param other the other value, not null
     * @return true if this is less than or equal the specified value, false
     * otherwise
     */
    fun isLessThanOrEqual(other: Num?): Boolean

    /**
     * Returns the minimum of this `num` and `other`.
     *
     * @param other value with which the minimum is to be computed
     * @return the `num` whose value is the lesser of this `num` and
     * `other`. If they are equal, method, `this` is returned.
     */
    fun min(other: Num): Num

    /**
     * Returns the maximum of this `num` and `other`.
     *
     * @param other value with which the maximum is to be computed
     * @return the `num` whose value is the greater of this `num` and
     * `other`. If they are equal, method, `this` is returned.
     */
    fun max(other: Num): Num

    /**
     * Returns the [Function] to convert a number instance into the
     * corresponding Num instance
     *
     * @return function which converts a number instance into the corresponding Num
     * instance
     */
    fun function(): Function<Number?, Num>

    /**
     * Transforms a [Number] into a new Num instance of this `Num`
     * implementation
     *
     * @param value the Number to transform
     * @return the corresponding Num implementation of the `value`
     */
    fun numOf(value: Number): Num {
        return function().apply(value)
    }

    /**
     * Transforms a [String] into a new Num instance of this with a precision
     * `Num` implementation
     *
     * @param value     the String to transform
     * @param precision the precision
     * @return the corresponding Num implementation of the `value`
     */
    fun numOf(value: String?, precision: Int): Num? {
        val mathContext = MathContext(precision, RoundingMode.HALF_UP)
        return this.numOf(BigDecimal(value, mathContext))
    }

    /**
     * Only for NaN this should be true
     *
     * @return false if this implementation is not NaN
     */
    val isNaN: Boolean
        get() = false

    /**
     * Converts this `num` to a `double`.
     *
     * @return this `num` converted to a `double`
     */
    fun doubleValue(): Double {
        return delegate!!.toDouble()
    }

    fun intValue(): Int {
        return delegate!!.toInt()
    }

    fun longValue(): Long {
        return delegate!!.toLong()
    }

    fun floatValue(): Float {
        return delegate!!.toFloat()
    }

    override fun hashCode(): Int
    override fun toString(): String

    /**
     * {@inheritDoc}
     */
    override fun equals(other: Any?): Boolean
}